cgsmiles/graph_layout_utils.py

```diff
@@ -0,0 +1,368 @@
+                    triplet = (remaining_nodes[0], common_node, remaining_nodes[1])
+                    triplets.add(triplet)
+    return triplets
+
+
+def assign_angles(graph, default_angle=120):
+    """
+    Assign angle values for all angles in a molecule
+    graph. We assume they are all 120 if atoms have
+    2 or 3 neighbors. Otherwise, they are skipped. On
+    top we take care that five membered rings and triangles
+    have the correct angle sum.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+
+    Returns
+    -------
+    tuple([[abc.hashable,abc.hashable,abc.hashable,int])
+        tuple with node keys corresponding to the angle
+        and the reference angle value
+    """
+    triplets = find_triplets(graph)
+    cycles = nx.cycle_basis(graph)
+    angles = []
+    for triplet in triplets:
+        # check if one is part of a ring
+        if any([triplet[0] in cycle for cycle in cycles]):
+            # find ring and check that all are part
+            for idx, cycle in enumerate(cycles):
+                if all([node in cycle for node in triplet]):
+                    angles.append((triplet[0],
+                                   triplet[1],
+                                   triplet[2],
+                                   180-360/len(cycle)))
+
+        if graph.edges[triplet[0], triplet[1]]['order'] > 2 or\
+            graph.edges[triplet[1], triplet[2]]['order'] > 2:
+            angles.append((triplet[0],
+                          triplet[1],
+                          triplet[2],
+                          180))
+
+        elif graph.degree(triplet[1]) in [2, 3]:
+            angles.append((triplet[0],
+                          triplet[1],
+                          triplet[2],
+                          default_angle))
+
+
+    return angles
+
+def assign_bonds(graph, pos, default_bond=None):
+    """
+    Iterate over all edges and compute the distance
+    between the edges defined by the `pos` coordinate
+    array.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+    pos: np.ndarray((n,2))
+
+    Returns
+    -------
+    tuple([abc.hashable, abc.hashable, float])
+        tuple of node keys and distance
+    """
+    bonds = []
+    for edge in graph.edges:
+        if default_bond:
+            dist = default_bond
+        else:
+            dist = np.linalg.norm(pos[edge[0]]-pos[edge[1]])
+        bonds.append((edge[0], edge[1], dist))
+    return bonds
+
+def assing_dihedral_angles(graph):
+    cis_trans = nx.get_node_attributes(graph, 'ez_isomer')
+    dihedrals = []
+    ref = {"cis": 0, "trans": 180}
+    for cis_trans_items in cis_trans.values():
+        for cis_trans_item in cis_trans_items:
+            dihedrals.append([*cis_trans_item[:4], ref[cis_trans_item[4]]])
+    return dihedrals
+
+def _bond_potential(atom_pos, ref):
+    """
+    Compute energy associated to the bonds.
+    """
+    energy = 100000*(ref-np.linalg.norm(atom_pos[1]-atom_pos[0]))**2
```

Comment:
`np.linalg.norm` is actually super slow (especially for single vectors, such as here). Doing something like `np.sqrt(((pos[edge[0]] - pos[edge[1]])**2).sum())` is faster. Only worth doing in inner loops though
```suggestion
    energy = 100000*(ref-np.sqrt(((pos[edge[0]] - pos[edge[1]])**2).sum()))**2
```

Reply:
🤣 we always have this discussion. But for all but very large matrices the performance difference is negligible. Somewhere in the polyply there is a reference to the StackOverflow post with the benchmarking